[4.0] Change ModuleChromes to JLayouts

Bakual opened the pull request:
Inspired on some ideas from Michael in https://github.com/joomla/joomla-cms/issues/20864 and some recent discussions about the Module Class behavior.
For those who aren't aware: The Module Chrome is the part around the actual module output. In the module settings, it is called "Module Style".

### Summary of Changes
This PR moves the module chromes (the code in that funny /templates/yourtemplate/html/modules.php file) to JLayouts.
The ones from the system templates are in the global /layouts/chromes/ folder.
The ones from the core templates are in the respective templates html/layouts/chromes/ folder.

#### Backward compatibility related
Support for chromes defined in modules.php is removed. Tenplates who wish to support both J3 and J4 can do so by adding the chromes as JLayouts and still keep the modules.php. In J3, the legacy one is used and in J4 the JLayouts are used.

The protected function `Joomla\CMS\Form\Field\ChromestyleField::getSystemTemplate()` is removed since it no longer is used.

Existing settings in modules for the module style keep working as the value stored in J3 and J4 is exactly the same.

### Testing Instructions
Set different module styles and test if the module is rendered with the expected chrome.
Test if you can add new layouts or override the existing chromes.

### Expected result
* Works from the JLayouts
* Custom chromes are possible
* Chromes are overrideable

### Actual result
* Works, but uses modules.php
* No customisation possible


### Documentation Changes Required
If module chromes are documented somewhere, that needs to be adjusted.

Commit:
Move Atum module chromes to JLayouts

Commit:
Move System module chromes to JLayouts

Commit:
Move Cassiopeia module chromes to JLayouts

Commit:
Remove reference to style "no". By default the system style "none" is used.

Commit:
Use JLayouts to render module chromes.

Commit:
Allow chromes from other templates

Commit:
Lookup JLayouts instead of module.php for available chromes. Need to lookup system JLayouts as well.

Commit:
Redordering Module Style above chrome related parameters

Commit:
Load global JLayouts as system styles

Commit:
Remove function getSystemTemplate and its call. Reorder code.

Commit:
Fixing "horz" chrome JLayout

Commit:
Simplify ModuleHelper

Bakual commented:
@mbabker I would be interested in your feedback since you also did some thinking on that.
@ciar4n Would love some feedback from you as well.

If that is approved, I'm thinking further to allow a XML file along the JLayout which would allow to define chrome specific parameters. But that will be a second PR and needs a bit more thoughts I guess.

dgrammatiko commented on layouts/chromes/tabs.php:
```diff
@@ -0,0 +1,25 @@
+
+if (!empty($content))
+{
+	echo HTMLHelper::_('tabs.panel', $module->title, 'module' . $module->id);
```
Do we still have this method in J4? AFAIK this was the mootools version which shouldn't be available anymore in J4

Bakual replied:
I haven't changed anything in the chrome itself. That's not the intent of this PR.
If the chrome is broken before, it will be broken afterwards as well 😄

dgrammatiko replied:
Well, the good part is that with this PR the issue is exposed, so either someone will have to redo that method (probably using the new Custom elements method) or this needs to be documented (I have no clue if that's already the case)
Obviously I'm not asking you to patch this in this PR, but maybe remove the broken parts?

mbabker replied:
This is a one-for-one swap of the `modChrome_tabs()` function to a layout file.  Leave it at that for now instead of introducing changes unrelated to the pull request (it also irks me when people try to slip CS fixes in untouched areas of a file in PRs but that's another story).

Bakual replied:
The CS stuff sneaked in by PhpStorm, I didn't notice them. Since I don't like such changes as well, I'm reverting those.

dgrammatiko commented on layouts/chromes/sliders.php:
```diff
@@ -0,0 +1,25 @@
+
+if (!empty($content))
+{
+	echo HTMLHelper::_('sliders.panel', $module->title, 'module' . $module->id);
```
Do we still have this method in J4? AFAIK this was the mootools version which shouldn't be available anymore in J4

Bakual replied:
Same as for the tabs one.

Commit:
Revert CS done by IDE automatically

Hackwar commented:
We indeed have to modernise the module chrome and 4.0 would be a good place to do this. However I would prefer if we would go a different route. I do kinda like that the chrome is not yet another anonymous file that no IDE knows how to treat, but a function (or method) that has a defined scope.

Thus I would propose to have a Joomla\CMS\Module\ModuleChrome class that implements a getChromeNames() method to have a list of available chromes in the current class and then a template specific class in a module.php or chrome.php in every template. That template class could extend from the system templates module chrome class which has a bunch of default chromes. The chromes are then just arbitrarily named methods of that class and the above mentioned method returns just the names of the functions that you can call. The ModuleChrome class could simply diff the methods of the base class and the inheriting class, while templates could implement their own code in case that isn't enough. 

This idea is more based on an uneasy feeling about creating a file for every chrome with all the overhead in the files, the difficulties for the IDEs and the overhead by using JLayout. I'm absolutely no fan of JLayout... But the simple fact that no IDE will be able to decipher the current scope of a JLayout file and all the copyright headers and merging variables into the current scope and all that, make me want to have a more object oriented approach. A chrome like `none` would be `public function none($module) {return $module->text;}` or something like that instead of the 40 lines that would be the prolog in every JLayout file...

wilsonge commented on layouts/chromes/horz.php:
```diff
@@ -0,0 +1,23 @@
+<?php
```
What is horz?

brianteeman replied:
isnt it a one for one swap for the existing chrome?

Bakual replied:
@wilsonge Ever looked at our existing module styles? 😄 
See https://github.com/joomla/joomla-cms/blob/d54007ccb54464b07092859a0c71aaa9c331dd42/templates/system/html/modules.php#L70-L83

Don't ask me who added that and used this 😄

wilsonge replied:
Ahh missed it in the frontend system module chrome. My bad.

ciar4n commented:
This makes a lot of sense. As you say, it brings chromes inline with how most other HTML is rendered. Although I've gotten used to it, initially the modules.php seemed detached to how HTML was generally rendered within Joomla. 

Been overridable gives plenty of obvious gains but from a template perspective, it allows changes to default chromes that users would be using (known or not). A module set as a tab for example (if one was to exist) could keep the same chrome settings between templates as the template would override that default chrome to match their own framework/personal code. Currently new chromes have to be created even thou differences may be subtle to the default. Considering that at some point it might be good to relook at the default chromes which in some cases are outdated and also how they are described. As Micheal pointed out in the original issue, the current method of selecting chromes is far from descriptive.

Bakual commented:
One thing I noticed is that the chrome jlayouts are not listed in the template override manager. That is because I've stored them in /layouts/chromes and the manager expects a folder more, eg /layouts/chromes/system.
Overrides work perfectly fine the way I currently did it.
Imho that limitation of the override manager is a bit stupid. Also the fact that it only creates overrides for the whole folder and not for a single file isn't that nice. So I'd rather see the override manager improved to be based on files, not folders which would fix this issue.

brianteeman commented:
> . Also the fact that it only creates overrides for the whole folder and not for a single file isn't that nice. 

I hit that one a lot yesterday.

ciar4n commented:
> If that is approved, I'm thinking further to allow a XML file along the JLayout which would allow to define chrome specific parameters. But that will be a second PR and needs a bit more thoughts I guess.

I guess the XML loaded would have to change dynamically depending on the 'Module Style' selected which AFAIK would be new territory for the UI. Loading each XML into a tab and then using 'showon'? Does 'showon' work on tabs?

ciar4n commented:
> This approach wouldn't be new. We use it for example for custom fields.

Of course... I'd forgotten about that.

In using 'showon' we would be relying on template devs to set this value in their custom XML which I guess would not be ideal. Otherwise, such settings will be displayed regardless of the 'Module Style' value. Not sure if there is a way around that.

Bakual commented:
See #23577 for an RFC about parameters

Quy commented on administrator/components/com_modules/forms/advanced.xml:
```diff
@@ -4,6 +4,12 @@
 			name="advanced">
 
+			<field
+					name="style"
```
```suggestion
				name="style"
```

Quy commented on administrator/components/com_modules/forms/advanced.xml:
```diff
@@ -4,6 +4,12 @@
 
+			<field
+					name="style"
+					type="chromestyle"
```
```suggestion
				type="chromestyle"
```

Quy commented on administrator/components/com_modules/forms/advanced.xml:
```diff
@@ -4,6 +4,12 @@
+			<field
+					name="style"
+					type="chromestyle"
+					label="COM_MODULES_FIELD_MODULE_STYLE_LABEL"
```
```suggestion
				label="COM_MODULES_FIELD_MODULE_STYLE_LABEL"
```

Quy commented on layouts/chromes/outline.php:
```diff
@@ -0,0 +1,31 @@
+?>
+<div class="mod-preview">
+	<div
+		class="mod-preview-info"><?php echo 'Position: ' . $module->position . ' [ Style: ' . $module->style . ']'; ?></div>
```
```suggestion
		class="mod-preview-info"><?php echo 'Position: ' . $module->position . ' [ Style: ' . $module->style . ' ]'; ?></div>
```

Commit:
CS

Co-Authored-By: Bakual <werbemails@bakual.ch>

Quy commented on administrator/templates/atum/html/layouts/chromes/body.php:
```diff
@@ -0,0 +1,36 @@
+	echo '<' . $moduleTag . ' class="' . $modulePos . ' card ' . htmlspecialchars($params->get('moduleclass_sfx')) . '">';
+	if ($module->showtitle && $headerClass !== 'card-title')
+	{
+		echo '<' . $headerTag . ' class="card-header' . $headerClass . '">' . $module->title . '</' . $headerTag . '>';
```
Add space after `card-header`?

Bakual replied:
Yeah, need to do that since we no longer use class suffixes.

Bakual replied:
Actually, I had that chrome completely wrong copied. I've redone it now correctly.
The space was missing in the original chrome as well, so I'm not changing that.

Quy commented on administrator/templates/atum/html/layouts/chromes/well.php:
```diff
@@ -0,0 +1,52 @@
+			<?php if ($canEdit) : ?>
+				<div class="module-actions">
+					<a href="<?php echo Route::_('index.php?option=com_modules&task=module.edit&id=' . (int) $module->id); ?>">
+						<span class="fa fa-cog"><span class="sr-only"><?php echo Text::_('JACTION_EDIT') . ' ' . $module->title; ?></span></span>
```
```suggestion
						<span class="fa fa-cog" aria-hidden="true"></span><span class="sr-only"><?php echo Text::_('JACTION_EDIT') . ' ' . $module->title; ?></span>
```

Bakual replied:
The chromes are a direct copy of the existing code. I'm not doing any changes here.

Quy commented on templates/cassiopeia/html/layouts/chromes/cardGrey.php:
```diff
@@ -0,0 +1,33 @@
+
+if ($module->content) : ?>
+	<<?php echo $moduleTag; ?> class="<?php echo $modulePos; ?> card card-grey<?php echo htmlspecialchars($params->get('moduleclass_sfx')); ?>">
+		<?php if ($module->showtitle and $headerClass !== 'card-title') : ?>
```
```suggestion
		<?php if ($module->showtitle && $headerClass !== 'card-title') : ?>
```

Bakual replied:
In Layouts we generally use "and" instead of "&&". So I'm not going to commit this change.

Commit:
Fix body chrome

Commit:
Adding space before moduleclasssuffix where it was in the original chrome.

wilsonge commented:
Can we get conflicts fixed here please. I'd like to get this in :)

Commit:
Merge branch '4.0-dev' into ChromeToJLayout

# Conflicts:
#	administrator/templates/atum/html/modules.php
#	administrator/templates/system/html/modules.php
#	libraries/src/Form/Field/ChromestyleField.php
#	templates/cassiopeia/html/modules.php
#	templates/system/html/modules.php

Commit:
Merge branch '4.0-dev' into ChromeToJLayout

Bakual commented:
Merged and conflicts solved.

alikon commented:
I have tested this item :white_check_mark: successfully on 9cfcc76b1dca151ce38572bb7f6f6ca75c1981d7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23570">issues.joomla.org/tracker/joomla-cms/23570</a>.</sub>

wilsonge committed:
Merge branch '4.0-dev' into ChromeToJLayout

wilsonge commented:
Dont' have a way of testing right now - but drone seems very unhappy here - can you check please

Bakual commented:
Drone is currently unhappy in the base branch and this PR was based on that "unhappy" branch. So I'd expect nothing else.
Check maintainer glip chat for details

Updating the branch once the base branch is fixed should solve it. 😄

Commit:
Merge branch '4.0-dev' into ChromeToJLayout

wilsonge commented:
I already did merge it in. And restarted drone like 4 times. So feel like it's a legit failure. Because it very rarely needs more than 2

Commit:
Check if chrome actually exists (returns something)

Commit:
There is no such chrome named "rounded" in J4

Bakual commented:
Ah, didn't see you already merged the branch. When I rebased the PR I saw it failing but already before the system tests even started. Now that this Javascript issue is solved, I saw the system test failure.

I now found the issue. It's actually because the login page requested a non-existing chrome ("rounded").
I've now added a fallback in case the chrome is missing plus removed the reference to the non-existing chrome.

wilsonge committed:
Merge branch '4.0-dev' into ChromeToJLayout

wilsonge committed:
Merge branch '4.0-dev' into ChromeToJLayout

wilsonge commented:
Thanks! This is a nice improvement!
